src/wallet/wallet.cpp

```diff
@@ -2739,8 +2747,16 @@ bool CWallet::CreateTransaction(const std::vector<CRecipient>& vecSend, CWalletT
                     return false;
                 }
 
-                LearnRelatedScripts(vchPubKey, g_change_type);
-                scriptChange = GetScriptForDestination(GetDestinationForKey(vchPubKey, g_change_type));
+                // If -changetype is not specified and g_address_type is not legacy,
+                // and any destination is P2WPKH or P2WSH, use P2WPKH for the change
+                // output.
+                OutputType change_type =
+                    g_change_type != OUTPUT_TYPE_NONE ? g_change_type :
+                    g_address_type != OUTPUT_TYPE_LEGACY && any_destination_native_segwit ? OUTPUT_TYPE_BECH32 :
+                    g_address_type;
+
+                LearnRelatedScripts(vchPubKey, change_type);
```

Comment:
I know this was here before but shouldn't this be called by CReserveKey only if the key isn't returned to the pool?